### DISCUSSION TITLE
About fallback models and environment files

### DISCUSSION THREAD

**Discussion by patillacode** [AntonOsika/gpt-engineer#135] [Ideas]
Hi everyone,

There have been numerous questions, issues and PRs (#14, #22, #29 , #41, #42,  #83, #92, #106, #144) regarding these matters, meaning there is definitely some interest and/or use cases. 

I think we should make a _final_ decision.
In order to do that I think a discussion would be valuable.

My two cents: if we adopted using a `.env` file, each of us could set the following:
```env
OPENAI_API_KEY=sk-xxxxxxxxxxxxxxxx
MODEL=gpt-4
...
```

Which for example, solves the fallback issue, where some people want to use `gpt-4` and others can't and have to use `gpt-3.5-turbo` (or `-16k` or `-16k-0613`)

I personally like the idea of each of us tweaking things like this instead of taking care of them in the code trying to fulfil everyone's need, each user can be responsible of their own _settings_, if you will.
So if I want to use a model that will soon be deprecated, like these new `-0613` which will eventually be deprecated, it is on me, to properly modify my `.env` or `settings.py`.

Also, many users have asked about the issue where they forgot to `export` their KEY and honestly, from my point of view, it is just annoying having to do the `export` every time I open a terminal to run GPT Engineer.
So, this solution would also solve that.

I am aware that @AntonOsika (repo owner) is [not convinced](https://github.com/AntonOsika/gpt-engineer/pull/29#issuecomment-1593514661) about this.

Just to be clear, maybe `.env` is not the way, maybe a proper `settings.py` file would do the trick. But I think we need _something_ to cover this cases.

I just thought we should discuss openly here and get some feedback from the community.

Please keep your comments on the matter, don't get too sidetracked if possible.

Let's discuss!

**Comment by djpapzin** [comment]
I support your point, I having to input my API key everytime I run is annoying

**Comment by Vandivier** [comment]
imo we should use .env for transparency, convenience, and security:
1. I like to stream programming. I can't use the cli like `OPENAI_KEY=foo python3 ...` on screen while streaming. it's more convenient and secure to have a .env file
2. A `.env.template` file makes it transparent what variables are expected.
3. A `.env.template` file is convenient because we can place default values where it makes sense

**Reply by jebarpg** [reply]
This is the way

**Comment by jnichols0** [comment]
I think having the desired model as an environment variable is a great idea and would not add to complexity or lines of code. Additionally  support for loading a local .env file would add value and could be implemented while respecting the goal of minimal lines of code. Reliable fallback logic has been added to support users without access to gpt-4, which is great because it should work "out of the box" for anyone with an openai account but I think it's reasonable that users shouldn't have to read and edit code to change something that could easily treated as a setting. 

Using [python-dotenv](https://pypi.org/project/python-dotenv/), adding just two lines of code to main.py :
```python
from dotenv import load_env
load_dotenv()
``` 
change the default arg in main() to
```python
model: str = os.getenv("GPT_ENGINEER_MODEL", "gpt-4")
```
would keep current behavior while adding support for configurable model via .env file or host env var GPT_ENGINEER_MODEL ( I think MODEL is too vague an env var name, might be safer to namespace it a bit). If no env var is set, it'd still try gpt-4, falling back on gpt-3.5-turbo. 

The README.md would need be updated to reflect the new env var and the support for a .env file.

For one dependency and two more total lines of code, this seems like a low cost high value feature to add.

**Comment by yaJannik** [comment]
+1 for me.

**Comment by yhyu13** [comment]
Just leave a trail on the previous pr on .env staff : https://github.com/AntonOsika/gpt-engineer/pull/29

Here is my fork on this project with .env support, just to keep it around as a reference in case anyone is interested: https://github.com/yhyu13/gpt-engineer/tree/local_dev

**Comment by RareMojo** [comment]
Big agree. This is how my fork works as it is more cost efficient to switch models depending on the task. Also, it's just more accessible in general, in my opinion.

+1

**Comment by AntonOsika** [comment]
Surprised by so many preferring env variables!

Why not just pass in `--model` parameter if you need it?

Is it because you run it from IDE / vscode?

**Reply by patillacode** [reply]
It's common practice and very handy especially for secrets!

**Comment by noxiouscardiumdimidium** [comment]
!cant use a gpu for gpt4all_ui, and i cant run textgen on an m40...but  if you can come up with ANY local llm integration capable of running on an m40 i can let you hit that with queries 24 hours a day... 

could you also implement support local llm hosting for 2 and only 2 backend tools.

A) official gpt4all_ui_v2.4.8 standalone one-click installer (cpu only, ggml, no-hassle in-app model downloads, buiit in openai_api calls, built in webserver api, built in localdoc interrogation) from your perspective the up shot to this is if you want to get a private vectorstore/database/datalake, people like me can download a pre currated data set of discrete vetted code snippets, load them in gpt4all ui localdocs, take whatever your current revision and improvement needs to meet your benchmark of having a beta_0 build compile the first "true" gpt-engineer_v1.0 ...

 you give me stable, local llm hosting integration - ill blast your vector store AROUND THE CLOCK... and with the webserver on one window process, and the main model window that can call openai calls  - if you make a burner account with a sharable key... maybe we could get some kind of distributed prompt serving going on ... is "Hyper-Modal" a word? it should be!
its got mpt-7b-instruct for spec (available for commercial use),  mpt-7b-chat for qa/clarify natural language breakdowns (available for commercial use), and replit-code-v1-3b for default, simple, and respec (available for commercial use)

B) OBVIOUSLY TEXTGEN!  xD

![Wed Jun 21 12:56:35 AM CDT 2023](https://github.com/AntonOsika/gpt-engineer/assets/115574840/866d1f29-7be1-4a91-aace-1aba9bcbb485)
![Wed Jun 21 01:17:43 AM CDT 2023](https://github.com/AntonOsika/gpt-engineer/assets/115574840/cc52282d-f979-4aae-8fed-3b5fae7ae8ea)
![Wed Jun 21 01:27:16 AM CDT 2023](https://github.com/AntonOsika/gpt-engineer/assets/115574840/7ca1aad1-5ffd-4926-8f74-16cb06036536)

**Reply by patillacode** [reply]
Sorry @noxiouscardiumdimidium 

This is not the place for this, this is a discussion about something else.

If you wish you can open an issue, please do it following the new issue template.

**Comment by jnichols0** [comment]
Given that gpt-engineer also accepts a model as an argument, there could be discrepancy between the two. E.g. supplying `gpt-3` via command line and `gpt-3.5-turbo` via the environment variables. I am wondering now if it could get confusing for the user to accept either or. 

Additionally, regarding model fall-back logic: my first contribution to this project was implementing a hard-coded fall-back to gpt-3.5-turbo since I don't have access to gpt-4 and I believe at the time there was no way to provide an alternative model to gpt-engineer w/o changing the code yourself. My intention was to make it work "out of the box" for whoever regardless of their API key. Now that  we can provide the model as a cli param, I think we should rethink that pattern. 

If a user pip installs gpt-engineer, sets up their openai key env var and runs `gpt-engineer projects/example` it should _just work_, try gpt-4 and fallback to gpt-3.5-turbo. But if the user asks for a specific model gpt-engineer can't do, e.g. `gpt-engineer projects/example --model no-such-model`, it should not just print a warning and revert to gpt-3.5-turbo and continue like that's just as good as what they asked for, because it might not be. I think we should only fall back from gpt-4 to gpt-3.5-turbo if the user has not specified any model explicitly.
